### PR TITLE
Metadata v2

### PR DESCRIPTION
**Pull request by ianhe8x**
I will continue @xlc 's work (#678) from here.
MetaNames are now converted into string. (https://gist.github.com/ianhe8x/6c6600a87eb724a795fa4dce83c69525)
the plan is 
1. in createType, all type class will be get from TypeRegistry.
2. in TypeRegistry, register the shipped types by default. change default export into function so it won't cause circular deps.
3. dynamically create type classes from metadataV2 and put them in TypeRegistry.

### CODE REVIEW

**Comment by jacogr** [comment]
@xlc & @ianhe8x 

Are you happy if I move the "primitive" types into `types/primitives`, these are things like `u32` & friends as defined as primitives here. (Truthfully the root is out of control, would at least want to start some basic cleanups so it is easier when know what we can drop as V2 is out there)

**Comment by ianhe8x** [comment]
I also think of something similar need to be done. And some day in the future we can remove the exports from index.ts, because now that every type comes from typeRegistry and I don't want user to have a wrong assumption.

**Comment by ianhe8x** [comment]
@jacogr in Type.ts#decodeType there're a few special treatments which seem no effect even against metadata v0. Is it safe to remove them? I will need to comment out `Type._removeGenerics()` at least.

**Comment by jacogr** [comment]
There are some, i.e. the spacing cleanups, that have been fixed in later v0 versions (since V0 was actually a version that evolved over time, latest is cleaner). 

It was tracked here - https://github.com/polkadot-js/api/issues/496, but that is basically non-existent in V2.

So for `removeGenerics`, we still (in V0/V1), get `<T as Trait>` which is stripped. (or `<TasTrait>` in _some_ less than optimal V0 versions...)

**Comment by ianhe8x** [comment]
I can create alias for those generics types in v0/v1 like ValidatorPrefs<Balance>. that would be a breaking change for some parachain devs, but anyway it will be after metadataV2 get merged into substrate.

**Comment by ianhe8x** [comment]
the following type implementation will be deprecated in V2 because from metadata I can't tell them from their real type and they don't have any special encoding/decoding.
```
Balance -> U128
BlockNumber -> U64
CodeHash -> H256
Hash -> H256
Gas -> U64
InherentOfflineReport => Null
LockPeriods ->  I8
Nonce -> U64
PropIndex -> U32
ProposalIndex -> U32
ReferendumIndex -> U32
VoteIndex -> VoteIndex
```

**Comment by jacogr** [comment]
Ok, it is hugely problematic to do just deprecation, let me take 2 examples -

`type Vote = i8`

In this case the API cares deeply to make it usable to actual devs, i.e. `-1` is true, it has extra fields to determine the yay/nay status, can be encoded with true/false. A simple `i8`, however it has logic from the runtime to make it sane.

`type Balance = u128`

The API doesn't care about encoding, however API users certainly does, For instance, apps uses this type info to display a different input field and (in the case of e.g. events) a different type of output.

Basically, the richness of the metadata cannot be lost - if it gets lost, we are making a huge step backwards.

**Comment by jacogr** [comment]
On generics - the `ValidatorPrefs<Balance>` doesn't add any addition info to the API user that a simple `ValidatorPrefs` does not do. I'd rather say the generic here is actually the implementation-specifics bleeding through. (Hence the API by default not using this currently, it doesn't add anything extra)

If needed for construction, it can probably be available - however in that case it is probably better keeping the type as it is (stripped) and having an extended type on, well, `Type` that carries this info - but only if REALLY needed for construction.

**Comment by xlc** [comment]
Sounds like I should add the original names back to make both available?
I really want to avoid that because that gives a hidden meaning of the type name, which is not documented anyway and will soon become maintenance burden. The handling of `BalanceOf` is an example, a refactor on the substrate side should not have any effect on polkadot.js side, which is not the case here.
A possibility is use new type approach similar to how I did to [AccountId here](https://github.com/paritytech/substrate/pull/1328/commits/2b871e71a9a714369f03755d07358ce0aa784614)

**Comment by jacogr** [comment]
Sure, I would argue that the metadata is actually not descriptive enough overall - i.e. https://github.com/paritytech/substrate/blob/master/srml/contract/src/lib.rs#L173 - that type would work well as `Code`, basically that would allow use of the API/UI/etc. to actually do validation, i.e. is this actually WASM?

So as we go down this road, it gets more and more complex - lately we have `BalanceOf` and `AssetOf` - from the pov of the API this is absolutely useless info and just adds unneeded complexity, basicqlly the implementation details are bleeding through.

In short - it is a fine line giving the correct information so users can actually make proper calls based on it and being all-over. So in the case of `Balance`, I would argue that `type Balance = u128` yields and entry in the metadata along the lines of (excuse my formatting here, it doesn't align with what is there, concepts are the same) -

```
{
  name: 'Balance',
  type: 'u128'
}
```

for `BalanceOf`, I'm guessing it is

```
{
  name: 'BalanceOf`
  type: 'Balance'
}
```

Which is probably still problematic in the `BalanceOf` case, since, once again, implementation details - the API should probably just want to treat this as a `Balance`

**Comment by xlc** [comment]
Another possibility I can think of is use explicit annotation similar to `#[compact]` to provide additional information that is not needed on substrate side, but very useful on the sdk side. Something like `#[metadata(Wasm)] code: Vec<u8>`

**Comment by ianhe8x** [comment]
There's a reason why Vote is not in the list, it has its own wrap type instead of i8. Though I understand your concern, as long as there's no convention that all types in substrate side must have a wrap type, we can't just remove support of mapping types by name.
On the other side, using type name and implementation together has its own downside. Raising more questions like which mode (by type name or by type signature) is the default mode? How to deal with name conflicts (one of the reason behind this PR)? These issues can be solved but at the cost of complexity in js side.
Personally I prefer the new type approach, though I don't have a say in this.

**Comment by gavofyork** [comment]
We're trying to take a program in a Turing-complete language and build a declarative type mapping from it that makes sense to a UI developer. It's just not going to happen. Basically, this is a nigh-on impossible problem and as such it's not going to have any clean solutions.

> This will require new explicit annotations on every usages of Balance

Introducing tags all over the Substrate codebase is not going to be acceptable.

In principle, I'm happy with a few well-placed tags in order to let a metadata compiler know that particular type definitions are useful to be placed in the metadata. e.g.

```
#[metadata]
type Balance = u128;
```

However, since these types may be used as generic parameters into other types (e.g. `BalanceOf`) or, worse, because of scoping, the same name may resolve to different types in different modules, it's far from a real solution. Indeed, the only "true" solution would be build Rust's type system in Javascript which obviously isn't going to happen.

Suppose, for example, you had a module that did this:

```
pub struct Record<T> { index: u32, data: T }
decl_storage! { trait Store for Module<T: Trait> as Broken {
  BalanceRecord: Record<BalanceOf<T>>;
  AccountIdRecord: Record<T::AccountId>;
}}
```

Now how could the metadata compiler represent that without exposing Rust's generic type system?

**Comment by xlc** [comment]
> Suppose, for example, you had a module that did this:
> 
> ```
> pub struct Record<T> { index: u32, data: T }
> decl_storage! { trait Store for Module<T: Trait> as Broken {
>   BalanceRecord: Record<BalanceOf<T>>;
>   AccountIdRecord: Record<T::AccountId>;
> }}
> ```
> Now how could the metadata compiler represent that without exposing Rust's generic type system?

My solution does expose Rust's generic type system and I think it works (there are rooms for improvements but I think it is good for now).

Take this example:

`Proposal` in treasury module

https://github.com/paritytech/substrate/blob/fbcdf4b36f1bec9455692b4f9d665139f1a97a94/srml/treasury/src/lib.rs#L131-L137

```
pub struct Proposal<AccountId, Balance> {
	proposer: AccountId,
	value: Balance,
	beneficiary: AccountId,
	bond: Balance,
}
```

https://github.com/paritytech/substrate/blob/fbcdf4b36f1bec9455692b4f9d665139f1a97a94/srml/treasury/src/lib.rs#L165

```
Proposals get(proposals): map ProposalIndex => Option<Proposal<T::AccountId, BalanceOf<T>>>;
```

Translate to this:

https://gist.github.com/xlc/3778f339afb9c1f4557ac9391cd14928#file-gistfile1-txt-L3669-L3682

```
Map {
    key: U32,
    value: CustomWithGenerics(
        "srml_treasury",
        "Proposal",
        [
            Custom(
                "sr_primitives",
                "AccountId"
            ),
            U128
        ]
    )
},
```

> Suppose you had two modules, each with an internal `Index` parameter:
> 
> ```
> module1 {
>   struct Index(u32);
>   decl_storage! { trait Store for Module<T: Trait> as Module1 {
>     CurrentIndex: Index;
>   }}
> }
> module1 {
>   struct Index(u64);
>   decl_storage! { trait Store for Module<T: Trait> as Module2 {
>     CurrentIndex: Index;
>   }}
> }
> ```
> Now how does the metadata deal with that?

The module name are included so it is possible to distinguish between name.

Please take at look of this for the reference output.

https://gist.github.com/xlc/3778f339afb9c1f4557ac9391cd14928#file-gistfile1-txt-L4562

**Comment by xlc** [comment]
Latest JOSN metadata: https://gist.github.com/xlc/c7f1125285dcd8330c195155fd0b5229

File: packages/types/src/Metadata/v2/MetadataRegistry.ts [hunk 1]
```diff
@@ -261,11 +261,7 @@ export class Metadata$H512 extends Null {
 }
 
 export class FieldName$Unnamed extends U16 {}
-export class FieldName$Named extends Text {
-  toString (): string {
```
Comment (xlc) [inline packages/types/src/Metadata/v2/MetadataRegistry.ts]: This is causing issue on codec because `encodedLength` depends on result of `toU8a` which depends on `toJSON` which depends on `toString`

**Comment by ianhe8x** [comment]
The code is now able to work with metadata v2. I would like to highlight the most important design decisions here for the convenience of discussion.
1. i don't want to make a v2 version of fromMetadata in type-storage and type-extrinsic, so I implement a toV0 for v2 as well. 
v2 types: `{"typeName": "T","displayName": "S" }` are disguised as `S|T`, and are decoded in createType.ts
2. Base.ts#init() is the place where metadataV2's extra type info be injected into TypeRegistry.
3. TypeDef in createType.ts now becomes
```
export type TypeDef = {
  displayType: DetailTypeDef,
  implType?: DetailTypeDef
};

export type DetailTypeDef = {
  info: TypeDefInfo,
  name?: string,
  type: string,
  sub?: DetailTypeDef | Array<DetailTypeDef>
};
```
to reflect the structure in MetadataV2, while `displayType` has a higher priority then `implType`.

Would like to hear from you @jacogr . and with a few more code clean up and some new tests, i will say this PR is ready for review and merge.

**Comment by jacogr** [comment]
I will -

- get it into apps (locally), testing against "current master", i.e. nothing breaks in current
- with the above, test against "v2 branch", i.e. everything still operates as it should with new

Judging by the v1 process, as long as the code is maintainable and understandable and it doesn't break the current running chains, could get this in before the v2 substrate merge (with the assumption that some changes may be made) - this helps in wider testing as well.

File: packages/types/src/Metadata/v2/static.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
...7401306e6f64655f72756e74696d65144576656e74012c73726d6c5f73797374656d1450686173650208384170706c7945787472696e736963000004000000123046696e616c697a6174696f6e01000001306e6f64655f72756e74696d65144576656e7402381873797374656d000004000000012c73726d6c5f73797374656d144576656e741c696e6469636573010004000000023073726d6c5f696e6469636573205261774576656e7408013473725f7072696d697469766573244163636f756e744964122062616c616e636573020004000000023473726d6c5f62616c616e636573205261774576656e7408013473725f7072696d697469766573244163636f756e744964161c73657373696f6e030004000000023073726d6c5f73657373696f6e205261774576656e7404141c7374616b696e67040004000000023073726d6c5f7374616b696e67205261774576656e740816013473725f7072696d697469766573244163636f756e7449642464656d6f6372616379050004000000023873726d6c5f64656d6f6372616379205261774576656e740816013473725f7072696d697469766573244163636f756e7449641c636f756e63696c060004000000024c73726d6c5f636f756e63696c3a3a7365617473205261774576656e7404013473725f7072696d697469766573244163636f756e74496438636f756e63696c5f766f74696e67070004000000025073726d6c5f636f756e63696c3a3a766f74696e67205261774576656e74041b3c636f756e63696c5f6d6f74696f6e73080004000000025473726d6c5f636f756e63696c3a3a6d6f74
```
Comment (amaurymartiny) [inline packages/types/src/Metadata/v2/static.ts]: I just built @xlc's branch on paritytech/substrate, and `state_getMetadata` gave me somthing starting with `0x1445`

File: packages/types/src/Metadata/v2/index.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+/**
+ * @name MetadataV2
+ * @description
+ * The runtime metadata as a decoded structure
+ */
+export default class MetadataV2 extends Struct implements MetadataInterface {
+  constructor (value?: any) {
+    super({
+      modules: Vector.with(MetadataModule),
+      module_path: Text,
```
[the remark this answers is not in the document]
Comment (xlc) [reply packages/types/src/Metadata/v2/index.ts]: `module_path` is generated from `module_path` macro in the substrate rust code in the runtime module. On substrate node it is `node_runtime`, on substrate template node it can be something else, on parachain it will be whatever module name the chain developer decides.

File: packages/types/src/Metadata/v2/latest.substrate.v2.json [hunk 4]
```diff
@@ -0,0 +1,5687 @@
...
+              "fallback": "0x0000000000000000000000000000000000000000000000000000000000000000",
+              "docs": []
+            },
+            {
+              "name": "ExtrinsicsRoot",
+              "modifier": "Default",
+              "type": {
+                "PlainType": {
+                  "typeName": "H256",
+                  "displayName": "Hash"
```
[the remark this answers is not in the document]
Comment (xlc) [reply packages/types/src/Metadata/v2/latest.substrate.v2.json]: I did the minimal amount of work to get it work (which is already a lot) so there are definitely rooms for optimizations.
The question is, is this a dealbreaker, or something can be improved later?

File: packages/types/src/Metadata/v2/static.ts (hunk 2 above)
Comment (ianhe8x) [reply packages/types/src/Metadata/v2/static.ts]: He did a rebase i think, will update it.

**Comment by ianhe8x** [comment]
updated per comment. and skip creating Tuple if length is 1.

File: packages/types/src/Metadata/v2/latest.substrate.v2.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/types/src/Metadata/v2/latest.substrate.v2.json]: i would suggest leave optimizations until we can get rid of legacy v0/v1. I did a lot of compromises in this PR because some codes in type-storage and type-extrinsic are bonded to v0's implementation. Otherwise I feel we're not able to improve more.
